# Data Consistency Checks Practice

## Mix type data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
path = r"C:\Users\anish\02-2025 Instacart Basket Analysis"

In [5]:
path

'C:\\Users\\anish\\02-2025 Instacart Basket Analysis'

In [7]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [12]:
# Creating a dataframe
df_test = pd.DataFrame()

In [14]:
# Create a mixted type column
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [18]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\anish\AppData\Local\Temp\ipykernel_29004\772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [22]:
# converting from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [26]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [28]:
df_nan = df_prods[df_prods['product_name'].isnull() == True ]

In [30]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [32]:
df_prods.shape

(49693, 5)

In [34]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [36]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [40]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [42]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [44]:
df_prods_clean.shape

(49677, 5)

In [46]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [48]:
df_prods_clean_no_dups.shape

(49672, 5)

## Task 4.5

In [53]:
# Question 2
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   orders_hour  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

After viewing the data, I believe there really isn't anything that is out of place or needs to be investigated further.

In [60]:
# Question 3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)  

C:\Users\anish\AppData\Local\Temp\ipykernel_29004\783386020.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\anish\AppData\Local\Temp\ipykernel_29004\783386020.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\anish\AppData\Local\Temp\ipykernel_29004\783386020.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\anish\AppData\Local\Temp\ipykernel_29004\783386020.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\anish\AppData\Local\Tem

In [62]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   orders_day_of_week      int64  
 6   orders_hour             int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


There is no mix type data in the orders data frame

In [65]:
# Question 4
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
orders_hour                    0
days_since_prior_order    206209
dtype: int64

There is no mix type data in the orders data frame

In [68]:
# Question 5
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

Unnamed: 0  order_id  eval_set  order_number  orders_day_of_week  \
count    206209.0  206209.0  206209.0      206209.0            206209.0   
mean          1.0       1.0       1.0           1.0                 1.0   
std           0.0       0.0       0.0           0.0                 0.0   
min           1.0       1.0       1.0           1.0                 1.0   
25%           1.0       1.0       1.0           1.0                 1.0   
50%           1.0       1.0       1.0           1.0                 1.0   
75%           1.0       1.0       1.0           1.0                 1.0   
max           1.0       1.0       1.0           1.0                 1.0   

       orders_hour  days_since_prior_order  
count     206209.0                206209.0  
mean           1.0                     0.0  
std            0.0                     0.0  
min            1.0                     0.0  
25%            1.0                     0.0  
50%            1.0                     0.0  
75%            1.0                     0.0  
max            1.0                     0.0

As we can see, there are 206209 missing values. From question 3, we know that there are 3421083 rows and 7 columns. The total number for unique user is 206209, which means each unique user_id has one missing value in the days_since_prior_order column, and that missing value belongs to the very first order of each unique user_id. All this makes sense as there is no value before each first order so we could leave these missing values alone.

#Question 6 
see explanation above ^

In [105]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         orders_hour  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0  
...              ...                     ...  
3421078           18                    29.0  
3421079           10                    30.0  
3421080           12                    18.0  
3421081           12                     7.0  
3421082           14                    30.0  

[3421083 rows x 7 columns]

In [ ]:
# Question 7-8

There are no duplicate values

In [108]:
# I'm going to create a new clean orders df by dropping the uneccessary unnamed column
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [110]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  orders_hour  \
0         2539329        1             1                   2            8   
1         2398795        1             2                   3            7   
2          473747        1             3                   3           12   
3         2254736        1             4                   4            7   
4          431534        1             5                   4           15   
...           ...      ...           ...                 ...          ...   
3421078   2266710   206209            10                   5           18   
3421079   1854736   206209            11                   4           10   
3421080    626363   206209            12                   1           12   
3421081   2977660   206209            13                   1           12   
3421082    272231   206209            14                   6           14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [112]:
df_prods_clean.to_csv(os.path.join(path,'Data','Prepared Data', 'product_checked.csv'))

In [114]:
df_ords_clean.to_csv(os.path.join(path,'Data','Prepared Data', 'orders_checked.csv'))